File: go/06-memoria/10-tecnicas/03-vector-store.go
// Vector store: SQL como fuente de verdad, embeddings como ranking semántico.
// Orquestación con degradación: si el índice falla, busca por texto.
//
// Cómo ejecutar: make go FILE=go/06-memoria/10-tecnicas/03-vector-store.go

package main

import (
	"fmt"
	"math"
	"sort"
	"strings"
	"time"
)

type Memoria struct {
	ID     string
	Texto  string
	Tipo   string
	Fuente string
	Creado int64
}

type ResultadoBusqueda struct {
	Memoria
	Score *float64
}

func mockEmbedding(texto string, dim int) []float64 {
	seed := uint64(0)
	for _, c := range texto {
		seed = seed*31 + uint64(c)
	}
	vec := make([]float64, dim)
	for i := range vec {
		seed = seed*1664525 + 1013904223
		vec[i] = float64(seed)/float64(math.MaxUint64)*2 - 1
	}
	norma := 0.0
	for _, v := range vec {
		norma += v * v
	}
	norma = math.Sqrt(norma)
	for i := range vec {
		vec[i] /= norma
	}
	return vec
}

func cosineSim(a, b []float64) float64 {
	sum := 0.0
	for i := range a {
		sum += a[i] * b[i]
	}
	return sum
}

type VectorStore struct {
	store        map[string]Memoria
	indice       map[string][]float64
	indiceActivo bool
	embDim       int
}

func NuevoVectorStore(dim int) *VectorStore {
	return &VectorStore{
		store:        make(map[string]Memoria),
		indice:       make(map[string][]float64),
		indiceActivo: true,
		embDim:       dim,
	}
}

func (vs *VectorStore) Insertar(id, texto, tipo, fuente string) {
	vs.store[id] = Memoria{ID: id, Texto: texto, Tipo: tipo, Fuente: fuente, Creado: time.Now().Unix()}
	vs.indice[id] = mockEmbedding(texto, vs.embDim)
}

func (vs *VectorStore) Buscar(query string, k int) []ResultadoBusqueda {
	if vs.indiceActivo && len(vs.indice) > 0 {
		return vs.buscarSemantico(query, k)
	}
	return vs.buscarTexto(query, k)
}

func (vs *VectorStore) buscarSemantico(query string, k int) []ResultadoBusqueda {
	qEmb := mockEmbedding(query, vs.embDim)
	type par struct {
		id    string
		score float64
	}
	pares := make([]par, 0, len(vs.indice))
	for id, emb := range vs.indice {
		pares = append(pares, par{id, cosineSim(qEmb, emb)})
	}
	sort.Slice(pares, func(i, j int) bool { return pares[i].score > pares[j].score })
	if k > len(pares) {
		k = len(pares)
	}
	resultados := make([]ResultadoBusqueda, k)
	for i, p := range pares[:k] {
		s := p.score
		resultados[i] = ResultadoBusqueda{Memoria: vs.store[p.id], Score: &s}
	}
	return resultados
}

func (vs *VectorStore) buscarTexto(query string, k int) []ResultadoBusqueda {
	tokens := strings.Fields(strings.ToLower(query))
	var resultados []ResultadoBusqueda
	for _, m := range vs.store {
		texto := strings.ToLower(m.Texto)
		for _, t := range tokens {
			if strings.Contains(texto, t) {
				resultados = append(resultados, ResultadoBusqueda{Memoria: m})
				break
			}
		}
		if len(resultados) >= k {
			break
		}
	}
	return resultados
}

func (vs *VectorStore) SimularFalloIndice() {
	vs.indiceActivo = false
	fmt.Println("  [simulación] índice vectorial desactivado")
}

func main() {
	store := NuevoVectorStore(64)

	recuerdos := [][3]string{
		{"m1", "Ana es la directora de producto de Acme Corp", "hecho"},
		{"m2", "El proyecto Pegasus tiene deadline en junio", "hecho"},
		{"m3", "El presupuesto del Q3 fue aprobado por Ana", "decision"},
		{"m4", "La integración con Stripe es parte de Pegasus", "hecho"},
		{"m5", "El equipo usa Python y Go como lenguajes principales", "hecho"},
		{"m6", "La reunión de lanzamiento fue el 15 de marzo", "evento"},
		{"m7", "El bug en auth afecta a usuarios admin", "hallazgo"},
		{"m8", "Ana aprobó el roadmap del Q4 en la reunión del viernes", "decision"},
	}

	for _, r := range recuerdos {
		store.Insertar(r[0], r[1], r[2], "")
	}
	fmt.Printf("Total recuerdos insertados: %d\n\n", len(store.store))

	fmt.Println("Búsqueda semántica: 'proyectos con presupuesto aprobado'")
	resultados := store.Buscar("proyectos con presupuesto aprobado", 3)
	for _, r := range resultados {
		score := "nil"
		if r.Score != nil {
			score = fmt.Sprintf("%.3f", *r.Score)
		}
		fmt.Printf("  [%s] %s\n", score, r.Texto)
	}

	fmt.Println()
	store.SimularFalloIndice()
	fmt.Println("Búsqueda con degradación: 'Ana proyecto'")
	resultadosFTS := store.Buscar("Ana proyecto", 3)
	for _, r := range resultadosFTS {
		fmt.Printf("  [texto] %s\n", r.Texto)
	}
}

File: typescript/06-memoria/10-tecnicas/03-vector-store.ts
// Vector store: SQL como fuente de verdad, embeddings como ranking semántico.
// Orquestación con degradación: si el índice vectorial falla, cae a búsqueda por texto.
//
// Cómo ejecutar: make ts SCRIPT=typescript/06-memoria/10-tecnicas/03-vector-store.ts

function mockEmbedding(texto: string, dim = 64): number[] {
  let seed = 0;
  for (let i = 0; i < texto.length; i++) seed = (seed * 31 + texto.charCodeAt(i)) >>> 0;

  const vec: number[] = [];
  for (let i = 0; i < dim; i++) {
    seed = (seed * 1664525 + 1013904223) >>> 0;
    vec.push(((seed >>> 0) / 0xffffffff) * 2 - 1);
  }
  const norma = Math.sqrt(vec.reduce((s, x) => s + x * x, 0));
  return vec.map((x) => x / norma);
}

function cosineSim(a: number[], b: number[]): number {
  return a.reduce((sum, ai, i) => sum + ai * b[i], 0);
}

interface Memoria {
  id: string;
  texto: string;
  tipo: string;
  fuente?: string;
  creado: number;
}

interface ResultadoBusqueda extends Memoria {
  score: number | null;
}

class VectorStore {
  private readonly store = new Map<string, Memoria>();
  private readonly indice = new Map<string, number[]>();
  private indiceActivo = true;

  insertar(id: string, texto: string, tipo = "hecho", fuente?: string): void {
    this.store.set(id, { id, texto, tipo, fuente, creado: Date.now() });
    this.indice.set(id, mockEmbedding(texto));
  }

  buscar(query: string, k = 5): ResultadoBusqueda[] {
    if (this.indiceActivo && this.indice.size > 0) {
      try {
        return this.buscarSemantico(query, k);
      } catch (e) {
        console.log(`  [degradación] índice vectorial falló → usando texto`);
        this.indiceActivo = false;
      }
    }
    return this.buscarTexto(query, k);
  }

  private buscarSemantico(query: string, k: number): ResultadoBusqueda[] {
    const qEmb = mockEmbedding(query);
    const scores = [...this.indice.entries()]
      .map(([id, emb]) => ({ id, score: cosineSim(qEmb, emb) }))
      .sort((a, b) => b.score - a.score)
      .slice(0, k);

    return scores.map(({ id, score }) => ({
      ...this.store.get(id)!,
      score,
    }));
  }

  private buscarTexto(query: string, k: number): ResultadoBusqueda[] {
    const tokens = query.toLowerCase().split(/\s+/);
    return [...this.store.values()]
      .filter((m) => tokens.some((t) => m.texto.toLowerCase().includes(t)))
      .slice(0, k)
      .map((m) => ({ ...m, score: null }));
  }

  simularFalloIndice(): void {
    this.indiceActivo = false;
    console.log("  [simulación] índice vectorial desactivado");
  }

  get total(): number {
    return this.store.size;
  }
}

// ── Demo ──────────────────────────────────────────────────────────────────

const store = new VectorStore();

const recuerdos: [string, string, string][] = [
  ["m1", "Ana es la directora de producto de Acme Corp", "hecho"],
  ["m2", "El proyecto Pegasus tiene deadline en junio", "hecho"],
  ["m3", "El presupuesto del Q3 fue aprobado por Ana", "decision"],
  ["m4", "La integración con Stripe es parte de Pegasus", "hecho"],
  ["m5", "El equipo usa Python y TypeScript como lenguajes principales", "hecho"],
  ["m6", "La reunión de lanzamiento fue el 15 de marzo", "evento"],
  ["m7", "El bug en el módulo de auth afecta a usuarios admin", "hallazgo"],
  ["m8", "Ana aprobó el roadmap del Q4 en la reunión del viernes", "decision"],
];

for (const [id, texto, tipo] of recuerdos) store.insertar(id, texto, tipo);
console.log(`Total recuerdos insertados: ${store.total}\n`);

console.log("Búsqueda semántica: 'proyectos con presupuesto aprobado'");
const resultados = store.buscar("proyectos con presupuesto aprobado", 3);
for (const r of resultados) {
  console.log(`  [${r.score?.toFixed(3)}] ${r.texto}`);
}

console.log();
store.simularFalloIndice();
console.log("Búsqueda con degradación: 'Ana proyecto'");
const resultadosFTS = store.buscar("Ana proyecto", 3);
for (const r of resultadosFTS) {
  console.log(`  [texto] ${r.texto}`);
}
